Automatically determine first weekday in wxCalendarCtrl

lanurmi opened the pull request:
Implement wxDateTime::GetFirstWeekDay(), and use is in wxCalendarCtrl to determine the preferred first weekday.

The implementation works on the three platforms it was tested on, but with platform-dependent code and a huge number of different locales it could potentially be used on, not every possible combination has been tested.

Also posted some information about this on wx-dev.

lanurmi committed:
Implement new static function wxDateTime::GetFirstWeekDay()

This function tries to determine the preferred first day of week to use in
calendars. The procedure for obtaining this information is highly
platform-dependent, and is not possible on all platforms; in that case
Sunday is used as the fallback value.

Implementations are included for MSW, OSX, and Linux.

lanurmi committed:
Automatically determine the first weekday for a calendar control

If neither wxCAL_SUNDAY_FIRST or wxCAL_MONDAY_FIRST was given, use
wxDateTime::GetFirstWeekDay() to automatically determine the preferred
day.

This changes the earlier default behaviour, which was to use Sunday
if not otherwise specified. However, the wxGTK native calendar control
ignored this behaviour anyway.

lanurmi committed:
Use automatically selected first weekday by default in calendar sample

Also create a wxLocale object with wxLANGUAGE_DEFAULT so that automatic
selection has a better chance of doing the right thing across different
platforms.

vadz commented on include/wx/datetime.h:
```diff
@@ -392,6 +392,9 @@ class WXDLLIMPEXP_BASE wxDateTime
     static bool IsLeapYear(int year = Inv_Year, Calendar cal = Gregorian);
 
+        // returns the first day of week based on locale and/or OS settings
+    static WeekDay GetFirstWeekDay(bool *success = NULL);
```
It doesn't _quite_ fit, but we already have `WeekFlags` enum, wouldn't it make sense to use it, rather than `WeekDay`, here? This would also make `success` parameter unnecessary...

vadz commented on interface/wx/datetime.h:
```diff
@@ -1479,6 +1479,15 @@ class wxDateTime
+	 @param success
+	     If given, the value will be set to true unless the first weekday
+	     could not be determined and the return value is a fallback value.
+    */
```
Could you please add `@since 3.1.1` here?

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
 }
 
+#ifdef _WIN32
+#include <wx/msw/registry.h>
```
Should use quotes instead of angle brackets in the code of the library itself.

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
+#include <wx/msw/registry.h>
+
+/* static */
+wxDateTime::WeekDay wxDateTime::GetFirstWeekDay(bool *success) {
```
Ideal would be to follow wx coding style, with braces on their own line (for functions, but also for `if` statements etc).

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
+
+/* static */
+wxDateTime::WeekDay wxDateTime::GetFirstWeekDay(bool *success) {
+    wxRegKey key("HKEY_CURRENT_USER\\Control Panel\\International");
```
I don't know if it really makes any difference, but all the other code uses `wxRegKey key(wxRegKey::HKCU, "...")` style ctor, so I'd prefer to use it here too for consistency.

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
+#elif defined(__linux__)
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
```
Is this really needed? It's supposed to be already defined by configure normally...

lanurmi replied:
It looks like defining `_GNU_SOURCE` is indeed not needed, I'll leave that out.

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
+wxDateTime::WeekDay wxDateTime::GetFirstWeekDay(bool *success) {
+    if (success)
+        *success = true;
+    return wxDateTime::WeekDay((*nl_langinfo(_NL_TIME_FIRST_WEEKDAY) - 1) % 7);
```
OTOH it might make sense to test for `_NL_TIME_FIRST_WEEKDAY` availability (using `#ifdef`). And we already have `HAVE_LANGINFO_H` for `nl_langinfo()` itself.

Like this, this code could work for non-Linux systems too if they use glibc (or implement this extension).

lanurmi replied:
I had subconsciously assumed that `_NL_TIME_FIRST_WEEKDAY` is not a macro despite the naming, and to my surprise I was right: it is an `enum` value, so an `#ifdef` would not work. I'll see if there is some other macro that would imply its presence. If not, maybe it would be best to check for it in the configure script.

vadz replied:
I had assumed it was a macro because I knew that other `nl_langinfo()` arguments were, but apparently glibc does this weird
```
enum {
  FOO,
#define FOO FOO
};
```
dance only for POSIX constants and not for the extensions... So yes, it does look like it'd be better to test for this in configure -- unless it's not supported by anything but glibc anyhow (and supported in all of its versions, or we could still have trouble with some old Linuxes).

vadz reviewed (commented):
Thanks for adding this function, it's indeed quite useful to have!

Could you please also update the existing `GetCountry() == USA` tests in `src/common/datetime.cpp` to use it?

vadz commented on src/common/calctrlcmn.cpp:
```diff
@@ -195,5 +195,18 @@ bool wxCalendarCtrlBase::SetHolidayAttrs()
     return true;
 }
 
+bool wxCalendarCtrlBase::WeekStartsOnMonday() const {
```
The same style comment as for the other commit.

vadz commented on src/common/calctrlcmn.cpp:
```diff
@@ -195,5 +195,18 @@ bool wxCalendarCtrlBase::SetHolidayAttrs()
+bool wxCalendarCtrlBase::WeekStartsOnMonday() const {
+    if (HasFlag(wxCAL_MONDAY_FIRST)) {
+        return true;
+    } else if(HasFlag(wxCAL_SUNDAY_FIRST)) {
```
We also use spaces more spaces, i.e. `if ( cond )` instead of `if(cond)`.

vadz commented on include/wx/calctrl.h:
```diff
@@ -26,12 +26,6 @@
-    // show Sunday as the first day of the week (default)
-    wxCAL_SUNDAY_FIRST               = 0x0000,
-
-    // show Monday as the first day of the week
```
Not sure why did the value of this one have to be changed, I'd leave it as is and just `SUNDAY_FIRST` to something non-zero.

vadz reviewed (approved):
Thanks!

lanurmi commented:
The API has now been changed to `bool GetFirstWeekDay(WeekDay *)`, the availability of `_NL_TIME_FIRST_WEEKDAY` is now checked in configure.

vadz commented on samples/calendar/calendar.cpp:
```diff
@@ -355,6 +363,10 @@ bool MyApp::OnInit()
 
+    // Locale affects on the language used in the calendar, and may affect
+    // on the first day of the week.
+    new wxLocale(wxLANGUAGE_DEFAULT);
```
This is a memory leak and while it's not exactly critical in a sample, I'd still prefer to avoid showing the wrong example in our code. Could you please just make it a member variable? Then we wouldn't need to allocate it on the heap at all, let alone free it.

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
     }
 }
 
+#ifdef _WIN32
```
We use `__WINDOWS__` for consistency.

vadz commented on src/common/datetime.cpp:
```diff
@@ -541,6 +541,52 @@ bool wxDateTime::IsLeapYear(int year, wxDateTime::Calendar cal)
+/* static */
+bool wxDateTime::GetFirstWeekDay(wxDateTime::WeekDay *firstDay)
+{
+    wxRegKey key(wxRegKey::HKCU, "Control Panel\\International");
```
I think there should be

    wxCHECK_MSG( firstDay, false, wxS("output parameter must be non-null") );

here (and, annoyingly, in the implementations of this function for the other platforms... until we want to split it into a public function and platform-dependent `DoGetFirstWeekDay()` which looks like an overkill).

vadz commented on src/common/datetime.cpp:
```diff
@@ -1723,7 +1769,9 @@ wxDateTime& wxDateTime::SetToWeekDayInSameWeek(WeekDay weekday, WeekFlags flags)
     if ( flags == Default_First )
     {
-        flags = GetCountry() == USA ? Sunday_First : Monday_First;
+        WeekDay firstDay;
```
Wouldn't it make sense to abstract this block in a function? E.g. Replace this entire `if` with

    UseEffectiveWeekDayFlags(flags);

vadz commented on include/wx/calctrl.h:
```diff
@@ -50,7 +47,10 @@ enum
+    wxCAL_SHOW_WEEK_NUMBERS          = 0x0040,
+
+	// show Sunday as the first day of the week (default)
+    wxCAL_SUNDAY_FIRST               = 0x0080
```
I don't remember what was the initial rationale for moving the enum elements around, but it doesn't seem necessary any more, so I'd like to avoid modifying this at all to avoid spurious diffs.

lanurmi replied:
The latest rationale was keeping the values in ascending order; I have now canceled moving the elements around, and just changed the value of `wxCAL_SUNDAY_FIRST`. If anyone adds a new element to the enum later, they'll need to notice that 0x80 is already reserved.

vadz commented on interface/wx/calctrl.h:
```diff
@@ -31,7 +28,10 @@ enum
+    wxCAL_SHOW_WEEK_NUMBERS          = 0x0040,
+
+    // show Sunday as the first day of the week (default)
+    wxCAL_SUNDAY_FIRST               = 0x0080
```
Same comment as for declarations.

vadz commented on interface/wx/datetime.h:
```diff
@@ -1484,6 +1484,18 @@ class wxDateTime
+         @param firstDay
+             The address of a WeekDay variable to which the first weekday will be
+			 assigned to.
+	     @return If the first day could not be determined, returns false,
```
Looks like a missing space of indent.

vadz reviewed (approved):
Thanks a lot, this looks very good now and I'm ready to apply it in the current state, but if you can improve the few (very) minor things I commented on, it would be even better. But if you're not going to have time to do it in near future, please just let me know and I'll merge it in its current state.

Thanks again!

lanurmi committed:
Refactor WeekFlags processing into UseEffectiveWeekDayFlags()

Use the new GetFirstWeekDay() function rather than GetCountry() == USA
to determine the first weekday.

vadz commented:
Thanks again for your work on this!
